### PR TITLE
Java Solution for Icy Perimeter

### PR DESCRIPTION
**Pull request by brad182**
I hope the comments aren't too extra, I just feel like I need to comment on many lines because the implementation can be a bit confusing

### CODE REVIEW

File: solutions/usaco-895.mdx [hunk 1]
```diff
@@ -2,16 +2,96 @@
...
+		for (int x = 1; x <= n; x++) {
+			for (int y = 1; y <= n; y++) {
+				if(numberedMatrix[x][y] == -1) { // not visited
+					floodfill(x, y, markNumber);
+					markNumber++;
+				}
+			}
+		}
+
+		int[] rowMovement = {0, 1, 0, -1}; // dx dy arrays
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:48]: make these constants

File: solutions/usaco-895.mdx [hunk 2]
```diff
@@ -2,16 +2,96 @@
...
+					numberedMatrix[x][y] = -1; // mark # as -1
+				}
+			}
+		}
+
+		// mark each connected component with a different number
+		int markNumber = 1;
+		for (int x = 1; x <= n; x++) {
+			for (int y = 1; y <= n; y++) {
+				if(numberedMatrix[x][y] == -1) { // not visited
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:41]: yeah two spaces after a line before you start a comment

File: solutions/usaco-895.mdx [hunk 3]
```diff
@@ -2,16 +2,96 @@
...
 <LanguageSection>
+	
+<JavaSection>
+	
+```java
+import java.io.*;
+import java.util.*;
 
-<CPPSection>
+public class IcyPerimeter {
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:21]:
```suggestion
public class Perimeter {
```
so it matches the actual problem name

File: solutions/usaco-895.mdx [hunk 4]
```diff
@@ -2,16 +2,96 @@
...
+	public static void main (String[] args) throws IOException {
+		Kattio io = new Kattio("perimeter");
+		n = io.nextInt();
+		numberedMatrix = new int[n + 2][n + 2];
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:25]: +2?
actually it's explained lower in the code
maybe add it here

File: solutions/usaco-895.mdx [hunk 5]
```diff
@@ -2,16 +2,96 @@
...
+			else if (areas[x] > largestArea) {
+				largestArea = areas[x];
+				perimeterOfLargestArea = perimeters[x];
+			}
+		}
+		io.println(largestArea + " " + perimeterOfLargestArea);
+		io.close();
+	}
+}
+
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:89]:
```suggestion
```

File: solutions/usaco-895.mdx [hunk 6]
```diff
@@ -2,16 +2,96 @@
...
+		// we do n + 2 because there is an extra outer ring for flood fill
+		for (int x = 0; x < (n + 2); x++) {
+			for (int y = 0; y < (n + 2); y++) {
+				// if part of a connected component
+				if (numberedMatrix[x][y] > 0) {
+					for (int i = 0; i < 4; i++) { // try all directions
+						// no need to check for out of bounds because of our outer ring
+						// if there is a zero, that means that we can add one to the perimeter
+						if(numberedMatrix[x + rowMovement[i]][y + columnMovement[i]] == 0) {
```
Comment (SansPapyrus683) [inline solutions/usaco-895.mdx:62]:
```suggestion
						if (numberedMatrix[x + rowMovement[i]][y + columnMovement[i]] == 0) {
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
too lazy to check actual substance

File: solutions/usaco-895.mdx (hunk 3 above)
Comment (brad182) [reply solutions/usaco-895.mdx:21]: the actual problem name is icy perimeter

File: solutions/usaco-895.mdx (hunk 1 above)
Comment (brad182) [reply solutions/usaco-895.mdx:48]: i dont think thats necessary its better to see the movements in the array than variables

**Review by dongliu0426 — APPROVED** [review]

**Review by maggie-j-liu — APPROVED** [review]

File: solutions/usaco-895.mdx (hunk 1 above)
Comment (SansPapyrus683) [reply solutions/usaco-895.mdx]: no i mean like
```java
public final int[] R_CHANGE = {0, 1, 0, -1};
```
or something

**Pull request merged**